### ISSUE TITLE
Rewrite tests to use a runtime, to make them less verbose

### ISSUE THREAD

**Issue by arduano** [urbas/nixrt#51]
Currently, the tests are implemented very manually, making it tedious to both write and edit them. Also, there are many duplicate tests between this repository and rix.

I propose making a single place for end-to-end integration tests, taking nix code and outputting the json result, and comparing them. This can either be done in this repo (by referencing the rix executable in the flake), or just writing these tests in rix. Maybe, it would be wise to make a monorepo (although keeping it modular, so people can grab just the runtime if they need still).

I remember you mentioned that you're not sure about the future of rix, was that purely because of c++? Like would you stay confident with it if there was a way to continue wrapping nix internals with rust?

**Comment by urbas** [comment]
Re end-to-end tests: Sounds very reasonable to me. But before starting this, maybe we should decide on the fate of `rix`.

Re C++: I am worrying about two things:
- using Rust will make it harder for us to use nix's C libraries
- using Rust will make it harder for nix to use rix's nix JS interpreter (if we want to try and replace the nix evaluation engine in nix with rix's)

I suspect it's possible to work around the two worries, but it will create extra work. I'm okay with continuing with Rust if you think the overhead of C-to-Rust bindings and also Rust-to-C bindings is not too bad.

Re future of rix: I'm thinking of narrowing down the scope of `rix` to just the evaluation bit and removing all the other goals. We can still expand the goals later, but until then I would clean the README.md and remove all the other Rust code that's not related to evaluation. What do you think?

**Comment by arduano** [comment]
C-to-Rust and Rust-to-C bindgs are very straightforward, the hard part is C++. However, there is a LOT of money currently in getting rust and c++ working together, funded my the likes of Microsoft and Google. I'm sure there's cleaner solutions coming out already.

Narrowing the scope also makes sense. Nix is very broad, and also has a lot of technical debt in its implementation of the concept, e.g. the `.narinfo` file format. Supporting everything faithfully would be a pain.

It should be possible for nix to transition to rust directly in the future anyway, kinda like other codebases such as `fish` have already done. Don't need to duplicate their work in a separate codebase imo.

**Comment by arduano** [comment]
@urbas should we move this repo into rix? Then they can invoke each other freely during tests. Or maybe just test in pure rust, without having jest at all?

**Comment by arduano** [comment]
I feel like it'd be a bit weird to have a cyclical reference between the two repos just for tests

**Comment by urbas** [comment]
Re repo into rix: Yes, let's do that.

Like you mentioned, let's start with just dropping this repo into a subdirectory and figure out how to improve and integrate the development environment later.

However, at a minimum we need to do the following:
- make sure GitHub actions still run for nixjs-rt
- rix's flake should use nixjs-rt's flake

I'll go ahead and do the move.

**Comment by urbas** [comment]
Btw, can you take a look at https://github.com/urbas/rix/pull/107 before merging the two repos?
